File: src/ros_study_cpp/src/my_action_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "ros_study_msgs/action/my_action.hpp"

class MyActionServer : public rclcpp::Node
{
public:
  using MyAction = ros_study_msgs::action::MyAction;
  using GoalHandleMyAction = rclcpp_action::ServerGoalHandle<MyAction>;

  MyActionServer()
  : Node("my_action_server")
  {
    // 액션 서버 생성
    action_server_ = rclcpp_action::create_server<MyAction>(
      this,
      "my_action",
      std::bind(&MyActionServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
      std::bind(&MyActionServer::handle_cancel, this, std::placeholders::_1),
      std::bind(&MyActionServer::handle_accepted, this, std::placeholders::_1)
    );
  }

private:
  rclcpp_action::Server<MyAction>::SharedPtr action_server_;

  // 목표 수락 여부 결정
  rclcpp_action::GoalResponse handle_goal(
    const rclcpp_action::GoalUUID &,
    std::shared_ptr<const MyAction::Goal> goal)
  {
    RCLCPP_INFO(this->get_logger(), "Received goal: %f", goal->go);
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  // 취소 요청 처리
  rclcpp_action::CancelResponse handle_cancel(
    const std::shared_ptr<GoalHandleMyAction>)
  {
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  // 목표가 수락되면 실행 스레드 시작
  void handle_accepted(const std::shared_ptr<GoalHandleMyAction> goal_handle)
  {
    std::thread{std::bind(&MyActionServer::execute, this, goal_handle)}.detach();
  }

  // 실제 목표 실행
  void execute(const std::shared_ptr<GoalHandleMyAction> goal_handle)
  {
    auto feedback = std::make_shared<MyAction::Feedback>();
    auto result = std::make_shared<MyAction::Result>();

    for (int i = 0; i < 5; ++i) {
      feedback->str.push_back("Step " + std::to_string(i));
      goal_handle->publish_feedback(feedback);
      RCLCPP_INFO(this->get_logger(), "Sending feedback: Step %d", i);
      rclcpp::sleep_for(std::chrono::seconds(1));
    }

    result->res = goal_handle->get_goal()->go * 2.0;
    goal_handle->succeed(result);
  }
};


int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MyActionServer>());
  rclcpp::shutdown();
  return 0;
}

File: src/ros_study_py/ros_study_py/my_action_server.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from ros_study_msgs.action import MyAction
from rclpy.qos import QoSProfile
import time

class MyActionServer(Node):
    def __init__(self):
        super().__init__('my_action_server')
        # 액션 서버 생성
        self._action_server = ActionServer(
            self,
            MyAction,
            'my_action',
            self.execute_callback
        )

    def execute_callback(self, goal_handle):
        self.get_logger().info(f'Received goal request: go={goal_handle.request.go}')

        feedback_msg = MyAction.Feedback()
        feedback_msg.str = []

        # 피드백 반복 전송 예시
        for i in range(5):
            feedback_msg.str.append(f"Step {i}")
            self.get_logger().info(f'Sending feedback: {feedback_msg.str}')
            goal_handle.publish_feedback(feedback_msg)
            time.sleep(1.0)

        # 목표 성공 처리
        goal_handle.succeed()

        # 최종 결과 생성
        result = MyAction.Result()
        result.res = goal_handle.request.go * 2.0
        return result


def main(args=None):
    rclpy.init(args=args)
    node = MyActionServer()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
